controller_manager/src/controller_manager.cpp

```diff
@@ -1311,6 +1296,18 @@ controller_interface::return_type ControllerManager::switch_controller(
     RCLCPP_DEBUG(get_logger(), " - %s", interface.c_str());
   }
 
+  // wait for deactivating async controllers to finish their current cycle
+  for (const auto & controller : deactivate_request_)
+  {
+    auto controller_it = std::find_if(
+      controllers.begin(), controllers.end(),
+      std::bind(controller_name_compare, std::placeholders::_1, controller));
+    if (controller_it != controllers.end())
+    {
+      controller_it->c->wait_for_trigger_update_to_finish();
```

Comment:
cannot this hang for an undetermined amount of time?

Reply:
Not really, this is just until the current cycle is finished. 

it might hang only when there is a crash or throw exception in another thread. That's why I added this one : https://github.com/ros-controls/realtime_tools/pull/172

Reply:
This is still a bit tricky, as it can easily disturb the update loop. for not is this fine. But I would love to see "async" deactivation of an async controller, or at least that we somehow earlier know what is happening.

The scenario I have in mind. My async controller is probably slow, so I would expect it to be blocking when writing for it to finish. which means that this would break the update loop here. I am not fully sure how to avoid this TBH, of how to manage it properly.

Reply:
Well but this is inside the nonRT loop, it is inside the switch_controllers method, so waiting there shouldn't be a problem

Reply:
What would be interesting is to skip the update trigger for the deactivated controllers to avoid it blocking